File: cuarray/samples/cpp/src/CuArray_argsort.cpp
#include <cuarray.h>
#include <iostream>

int main() {
    std::cout
            << "Running "
            << __FILE__
            << std::endl;

/* Creates a new float CuArray instance */
    CuArray<float> *cuArray = new CuArray<float>();

/* Initialize the CuArray with 300 rows and 300 columns */
    auto rows = 300;
    auto cols = 300;
    cuArray->init(rows,
                  cols);

/* Fill the CuArray with random values */
    for (int i = 0; i < cuArray->m(); i++) {
        for (int j = 0; j < cuArray->n(); j++) {
            cuArray->host()[i * cuArray->n() + j] =
                    static_cast<float>(rand() / (float) RAND_MAX);
        }
    }

/* Create a new CuArray with indices that sort the 8th row
 * of the original CuArray.*/
    auto cuArrayRowIndex = 7;
    auto sortedIndicesCuArray = cuArray->argsort(cuArrayRowIndex);

/* Create a new CuArray containing sorted data from the 8th row
 * of the original CuArray.*/
    auto sortedCuArray = cuArray->sort(cuArrayRowIndex);

/* Print the sorted CuArray and the corresponding values from the
 * original CuArray using the sortedIndicesCuArray.*/
    for (int j = 0; j < sortedCuArray->n(); j++) {
        auto sortedIndex = sortedIndicesCuArray->get(0,
                                                     j);
        auto sortedValue = sortedCuArray->get(0,
                                              j);
        auto sortedValueFromOriginalCuArray =
                cuArray->get(sortedIndex,
                             cuArrayRowIndex);
        std::cout
                << sortedIndex
                << " "
                << sortedValue
                << " "
                << sortedValueFromOriginalCuArray
                << std::endl;
    }

/* Cleanup time. */
    delete cuArray;
    delete sortedCuArray;
    delete sortedIndicesCuArray;
    return 0;
}

File: cuarray/samples/python/CuArray_argsort.py
import numpy as np

"""
Always precede CuArray with the data type
Here we are importing the CuArray int and float templates 
"""
from cuarray import FloatCuArray

print("Running", __file__)

"""
Create a new float CuArray instance 
"""
float_cuarray = FloatCuArray()

"""
Create a random float32 numpy array with 10 rows
and 10 columns
"""
numpy_array = np.random.rand(10, 10).astype(np.float32)

"""Load the numpy array into the CuArray"""
float_cuarray.fromNumpy2D(numpy_array)

"""
Perform a descending sort on 
the 8th row of float_cuarray
"""
sorted_cuarray = float_cuarray.sort(7)

"""
Get the indices that sort the 8th row of float_cuarray
"""
argsort_cuarray = float_cuarray.argsort(7)

"""
Print the sorted 8th row of float_cuarray using 
sorted_cuarray and argsort_cuarray indices
"""
for _ in range(10):
    sort_idx = argsort_cuarray[0][_]
    print(
        sorted_cuarray[0][_],
        float_cuarray[7][sort_idx]
    )